### ISSUE TITLE
🐞 Parent directory is uploaded when working in a module subdir

### ISSUE THREAD

**Issue by gerhard** [dagger/dagger#8395]
### What is the issue?

When running a module from a subdirectory, the parent directory gets uploaded to the Engine:

```console
# ...
cache request: upload /var/folders/z5/n5y08g4j073_ld2d9019j3_80000gn/T/tmp.s1fViQInu2/dagger from 45fhofmgv7lscg6ac23c50pz3
upload /var/folders/z5/n5y08g4j073_ld2d9019j3_80000gn/T/tmp.s1fViQInu2/dagger from 45fhofmgv7lscg6ac23c50pz3
copy upload /var/folders/z5/n5y08g4j073_ld2d9019j3_80000gn/T/tmp.s1fViQInu2/dagger from 45fhofmgv7lscg6ac23c50pz3
# ...
```

As discovered with @aluzzardi earlier today.

See the steps on how to reproduce & the Dagger Cloud Trace link. FWIW:

<img width="1552" alt="image" src="https://github.com/user-attachments/assets/0b0cdda9-f539-4933-93a6-5bb584b50b71">

### Dagger version

dagger v0.12.7 (registry.dagger.io/engine:v0.12.7) darwin/amd64

### Steps to reproduce

```console
cd $(mktemp -d)

git clone https://github.com/dagger/dagger
cd dagger

mkdir foo
cd foo
dagger init --sdk=go --name=foo -vvv
```

### Log output

https://dagger.cloud/dagger/traces/5d85d7b1d5483304f37b6c8dfc3db8e2

**Comment by helderco** [comment]
Uhh.. this explains a lot actually!

**Comment by helderco** [comment]
Wait, taking a closer look. It seems that it's loading the context directory, not necessarily the parent directory right? That's expected.

**Comment by gerhard** [comment]
I'm running the command in `dagger/foo`. That makes me think that the context directory is `foo`. There is no reason for it to go up a level - there is no configuration in the default init module that would make it behave this way.

Did you try to follow the steps to reproduce?

**Comment by helderco** [comment]
A module has 3 directories (ignore the new "context directories" feature as that's something different):
- **context directory**: git root (where `.git` is) or the _root directory_ if not in a git repo
- **root directory**: the directory where  `dagger.json` is
- **source directory**: the directory referenced in `"source"` in `dagger.json`

So in effect, if you're in the root of the dagger/dagger repo and `dagger init --sdk=go foo`, that'll load all the Go code in the repo along with that  `foo` module. Even in other sibling directories.

**Comment by helderco** [comment]
I followed your steps to reproduce and this is what I see that is being uploaded in `resolveFromCaller`: https://app.warp.dev/block/8wnHMlOFRKJWTysCpr696N

Notice the size: 4493546. Isn't that just ≈4MB though? Perhaps it's the sheer quantity of files that's at fault more than the size.

**Comment by helderco** [comment]
Note that this depends on what you have locally. For example I have `sdk/php/vendor` and that's the biggest contributor. If you have other modules in the repo with generated code, that should be loaded as well.

**Comment by helderco** [comment]
On a clean clone I get ≈6MiB of disk usage, with 899 items: https://app.warp.dev/block/DlM5Ecv0hOPWkFLDvdVy1r

**Comment by gerhard** [comment]
Why is the directory where `.git` is getting uploaded?

This is the unexpected part for me, even though it seems that it's working as expected.

**Comment by helderco** [comment]
I don't think `.git` is being uploaded (as it's excluded by default), it's the directory where `.git` is, i.e., the repo root.

**Comment by gerhard** [comment]
Yes: why is the context directory (where `.git` is) getting uploaded?

**Comment by jedevc** [comment]
I think this is a somewhat known issue? `dagger init` uploads parent directories - up to the root, this *was* so it could detect parent `go.mod`s and similar. I'm sure we've discussed this before (I seem to remember with @jpadams, there was a conversation around daggerizing existing repos with this).

It's not fully necessary to do this in most cases, and we could probably do much better with the context dir primitives we now have.

---

For running a module, I'm not sure. Can you share the module that's in the subdir that you're hitting the issue with?

**Comment by gerhard** [comment]
Yes, this definitely doesn't seem right.

The module is: `dagger init --sdk=go --name=foo -vvv`

**Comment by jedevc** [comment]
@sipsma I think with the refactorings to the module loading system these issues should be resolved now right? We should only be uploading the files needed for the module.
